src/tree/zk.rs

```diff
@@ -15,65 +15,237 @@ pub fn merkle_opening<const DEPTH: usize>(
     branch: &PoseidonBranch<DEPTH>,
     leaf: Witness,
 ) -> Witness {
-    let zero = TurboComposer::constant_zero();
+    // Generate a permutation container
+    let mut container = [TurboComposer::constant_zero(); dusk_hades::WIDTH];
 
-    let mut base = true;
-    let mut root = zero;
+    // Initialize the root with the leaf hash
+    let mut root = leaf;
 
-    // Generate a permutation container
-    let mut perm = [zero; dusk_hades::WIDTH];
+    // Recalculate the root for the given branch
+    for depth in 0..DEPTH {
+        let level = branch.as_ref()[depth];
 
-    // For every level, replace the level offset with needle,
-    // permutate the level and set the needle to the next level
-    // to the poseidon result of the permutation
-    branch.as_ref().iter().for_each(|level| {
         // Create the bits representation of the offset as witness
+        // and make sure that offset points to a hash in the level
         let offset_flag = level.offset_flag();
-
-        let mut sum = zero;
-        let mut offset_bits = [zero; dusk_hades::WIDTH - 1];
+        let mut sum = TurboComposer::constant_zero();
```

Comment:
The original purpose of this block was to validate that we have only zeroes/ones, and that we have one, and only one `1`; the rest being `0`.

It was written before we had [component_boolean](https://docs.rs/dusk-plonk/0.12.0/dusk_plonk/constraint_system/struct.TurboComposer.html#method.component_boolean) available, which now completely supersede this implementation. What we need is to run the boolean component for every of these bits (so we guarantee they are either `0` or `1`, and check that they add to `1` (so we guarantee that we have one, and only one `1`).

Reply:
Good point! I'll address this in a new issue #188